_posts/you-wanna-test-http.md

```diff
@@ -0,0 +1,215 @@
+
+I'm quoting from the above-linked documentation page. It sounds like the purpose of the `TEntryPoint` type is to set up the dependency context and not necessarily to link to a startup class. It turns out that we can write the following:
+
+```csharp
+class WebApplicationFactoryWithHost<TEntryPoint> :
+	 WebApplicationFactory<TEntryPoint>
+	 where TEntryPoint : class
+{
+    Action<IServiceCollection> configureServices;
+    Action<IApplicationBuilder> configure;
+    string[] args;
+
+    public Action<IHostBuilder> HostBuilderCustomization { get; set; }
+    public Action<IWebHostBuilder> WebHostBuilderCustomization { get; set; }
+
+    public WebApplicationFactoryWithHost(Action<IServiceCollection> configureServices, Action<IApplicationBuilder> configure, string[] args = null)
+    {
+        this.configureServices = configureServices;
+        this.configure = configure;
+        this.args = args ?? new string[0];
+    }
+
+    protected override IHostBuilder CreateHostBuilder()
+    {
+        var hostBuilder = Host.CreateDefaultBuilder(args);
+        hostBuilder.ConfigureWebHostDefaults(webBuilder =>
+        {
+            webBuilder.ConfigureServices(configureServices);
+            webBuilder.Configure(configure);
+
+            WebHostBuilderCustomization?.Invoke(webBuilder);
+        });
+
+        HostBuilderCustomization?.Invoke(hostBuilder);
+
+        return hostBuilder;
+    }
+}
+```
+
+I feel that what the above class is doing is pretty straightforward and doesn't need any comment. The presented `WebApplicationFactoryWithHost` is nothing more than a sample. There are many enhancements that we can apply. It just works for my use case.
+
+Let's have a look at the usage:
+
+```csharp
+[Fact]
+async Task SampleMiddlewareTest()
+{
+   var factory = new WebApplicationFactoryWithHost<Dummy>
+   (
+      configureServices: services =>
+      {
+         //configure services as needed
+      }
+      configure: app =>
+      {
+         app.UseRequestCulture();
+         //rest of the required app configuration here
+      }
+   );
+
+   var client = factory.CreateClient();
+   var response = await client.GetAsync("/sample/1?culture=it-IT");
+
+   Assert.True(response.IsSuccessStatusCode);
+   //more assertions to validate the culture settings
+}
+```
+
+Cool, isn't it? We're self-hosting the whole web application in the test. We control the services customization and the application set up when started. Finally, we can exercise it like if it was an actual web application.
+
+> The `Dummy` class is nothing else that a marker/empty class defined in the test project. It can be whatever type in the test project; the runtime will use it to identify the assembly to create the dependency context to load dependencies.
```

Comment:
```suggestion
> The `Dummy` class is simply a marker/empty class defined in the test project. It can be whatever type in the test project; the runtime will use it to identify the assembly to create the dependency context to load dependencies.
```